chain/postgresdb/pgdb/pgdb.go

```diff
@@ -0,0 +1,441 @@
+		ID: c.store.beaconID,
+	}
+
+	var dbBeacons []dbBeacon
+	if err := database.NamedQuerySlice(ctx, c.store.log, c.store.db, query, data, &dbBeacons); err != nil {
+		return nil, err
+	}
+
+	if len(dbBeacons) == 0 {
+		return nil, chainerrors.ErrNoBeaconStored
+	}
+
+	return toChainBeacon(dbBeacons[0]), nil
+}
+
+// Next returns the next beacon from the configured beacon table.
+func (c *cursor) Next(ctx context.Context) (*chain.Beacon, error) {
+	defer func() {
+		c.pos++
+	}()
+
+	const query = `
+	SELECT
+		round, signature, previous_sig
+	FROM
+		(SELECT
+			round, signature, previous_sig
+		FROM
+			beacon_details
+		WHERE
+			beacon_id = :id
+		ORDER BY
+			round ASC OFFSET :offset
+		LIMIT 1) AS T
+	WHERE
+		round IS NOT NULL`
+
+	data := struct {
+		ID     int    `db:"id"`
+		Offset uint64 `db:"offset"`
+	}{
+		ID:     c.store.beaconID,
+		Offset: c.pos + 1,
+	}
+
+	var dbBeacons []dbBeacon
+	if err := database.NamedQuerySlice(ctx, c.store.log, c.store.db, query, data, &dbBeacons); err != nil {
+		return nil, err
+	}
+
+	if len(dbBeacons) == 0 {
+		return nil, chainerrors.ErrNoBeaconStored
+	}
+
+	return toChainBeacon(dbBeacons[0]), nil
+}
+
+// Seek searches the beacon table for the specified round
+func (c *cursor) Seek(ctx context.Context, round uint64) (*chain.Beacon, error) {
+	const query = `
+	SELECT
+		round, signature, previous_sig
+	FROM
+		(SELECT
+			round, signature, previous_sig
+		FROM
+			beacon_details
+		WHERE
+			beacon_id = :id AND
+			round = :round
+		LIMIT 1) AS T
+	WHERE
+		round IS NOT NULL`
+
+	data := struct {
+		ID    int    `db:"id"`
+		Round uint64 `db:"round"`
+	}{
+		ID:    c.store.beaconID,
+		Round: round,
+	}
+
+	var dbBeacon dbBeacon
+	if err := database.NamedQueryStruct(ctx, c.store.log, c.store.db, query, data, &dbBeacon); err != nil {
+		return nil, chainerrors.ErrNoBeaconStored
+	}
+
+	if err := c.seekPosition(ctx, round); err != nil {
+		return nil, err
+	}
+
+	return toChainBeacon(dbBeacon), nil
+}
+
+// Last returns the last beacon from the configured beacon table.
+func (c *cursor) Last(ctx context.Context) (*chain.Beacon, error) {
+	const query = `
+	SELECT
+		round, signature, previous_sig
+	FROM
+		(SELECT
+			round, signature, previous_sig
+		FROM
+			beacon_details
+		WHERE
+			beacon_id = :id
+		ORDER BY
+			round DESC
+		LIMIT 1) AS T
+	WHERE
+		round IS NOT NULL`
+
+	data := struct {
+		ID int `db:"id"`
+	}{
+		ID: c.store.beaconID,
+	}
+
+	var dbBeacons []dbBeacon
+	if err := database.NamedQuerySlice(ctx, c.store.log, c.store.db, query, data, &dbBeacons); err != nil {
+		return nil, err
+	}
+
+	if len(dbBeacons) == 0 {
+		return nil, chainerrors.ErrNoBeaconStored
+	}
+
+	if err := c.seekPosition(ctx, dbBeacons[0].Round); err != nil {
+		return nil, err
+	}
+
+	return toChainBeacon(dbBeacons[0]), nil
+}
+
+// seekPosition updates the cursor position in the database for the next operation to work
+func (c *cursor) seekPosition(ctx context.Context, round uint64) error {
+	const query = `
```

Comment:
This will be relatively expensive.
Also, I can try to rewrite it as a more generic-SQL query, if desired.
Any thoughts on doing it?

Reply:
Fixed in https://github.com/drand/drand/pull/1088/commits/81dcc0e535d04b5ccbe3cdcc8e566e1c618d963b